### ISSUE TITLE
훈련 세트와 테스트 세트

### ISSUE THREAD

**Issue by wongakim-99** [wongakim-99/ml-dl-notes#2]
지도 학습과 비지도 학습의 차이를 공부한다. 모델을 훈련시키는 훈련 세트와 모델을 평가하기 위한 테스트 세트로 데이터를 나눠서 학습해보자.

**Comment by wongakim-99** [comment]
다음 문장에서 이상한 점을 찾아보자.

"도미 35마리와 빙어 14마리를 모두 저장하고 맞추는 거라면 100%를 달성하는 게 당연하지 않는가? 어떤 것이 도미이고 빙어인지 알고 있는데 맞추지 못하는 것이 이상하다."

**Comment by wongakim-99** [comment]
## 지도 학습과 비지도 학습

머신러닝 알고리즘은 크게 지도 학습(supervised learning) 과 비지도 학습(unsupervised learning)으로 나눌 수 있다. 지도 학습 알고리즘은 훈련하기 위한 데이터와 정답이 필요하다.

이전 이슈에서 보았던 도미와 빙어의 예를 보면 생선의 길이와 무게를 알고리즘에 사용했다. 이 경우 정답은 도미인지 아닌지 여부이다. 지도 학습에서는데이터와 정답을 **입력(input) 과 타깃(target)**이라고 하고, 이 둘을 합쳐 **훈련 데이터 training data** 라고 부른다. (지도 학습 알고리즘은 입력(데이터)과 타깃(정답)으로 이뤄진 훈련 데이터가 필요하다.)

그리고 앞서 언급했듯이 입력으로 사용된 길이와 무게를 **특성(feature)** 이라고 한다. 도미와 빙어를 구분하는 데 사용한 길이와 무게가 특성인 것이다.

- **입력, 타깃, 특성, 훈련 데이터 (잘 기억하기)** 

![image](https://github.com/user-attachments/assets/43018ba3-5196-4b04-8602-ebbdc1ac10d5)

- 지도 학습 : 정답(타깃)이 있으니 알고리즘이 정답을 맞히는 것을 학습한다. 예를 들어 도미인지 빙어인지 구분하는 것이다.
- 비지도 학습 : 타깃 없이 입력 데이터만 사용한다. 이런 종류의 알고리즘은 정답을 사용하지 않으므로 무언가를 맞힐 수가 없다.

※ 머신러닝 알고리즘은 지도 학습, 비지도 학습 외에 강화 학습(reinforcement learning)을 또 다른 종류로 크게 분류한다. 강화 학습 알고리즘은 타깃이 아니라 알고리즘이 행동한 결과로 얻은 보상을 사용해 학습된다.

![image](https://github.com/user-attachments/assets/bf3dda2f-a19a-4b87-80d1-588a8a542b46)

1번 이슈에서 도미와 빙어를 구분하기 위해 사용한 KNN 알고리즘은 입력 데이터와 타깃(정답)을 사용했으므로 당연히 지도 학습 알고리즘이다. 이 알고리즘을 훈련하여 생선이 도미인지 아닌지를 판별하고, 이 모델이 훈련 데이터에서 도미를 100% 완벽하게 판별했다. 모든 것이 잘 된 것 같은데 무엇이 문제일까??

**Comment by wongakim-99** [comment]
## 훈련 세트와 테스트 세트

기말고사를 보기 전에 출제될 시험 문제와 정답을 미리 알려주고 시험을 본다면 어떨까? 바보가 아닌 이상 답을 외우는 최소한의 노력만 한다면 만점은 가능할 것이다.

머신러닝도 이와 마찬가지이다. 도미와 빙어의 데이터와 타깃을 주고 훈련한 다음, 같은 데이터로 테스트한다면 모두 맞히는 것이 당연하다. 연습문제와 시험 문제가 달라야 올바르게 학생의 능력을 평가할 수 있듯이 머신러닝 알고리즘의 성능을 제대로 평가하려면 훈련 데이터와 평가에 사용할 데이터가 각각 달라야 한다.

이렇게 하는 가장 간단한 방법은 평가를 위해 또 다른 데이터를 준비하거나 이미 준비된 데이터 중에서 일부를 떼어 내어 활용하는 것이다. 일반적으로 후자의 경우가 많다.

평가에 사용하는 데이터를 **테스트 세트(test set)** , 훈련에 사용되는 데이터를 **훈련 세트(train set)** 라고 부른다.
정리해 보자. 우리는 모델을 훈련하기 위한 데이터를 사용해 이 모델의 정확도를 100%라고 평가했다. 하지만 훈련에 사용한 데이터로 모델을 평가하는 것은 적절하지 않다. 훈련할 때 사용하지 않은 데이터로 평가해야 한다. 이를 위해 훈련 데이터에서 일부를 떼어 내어 테스트 세트로 사용하겠다.

```python
fish_length = [25.4, 26.3, 26.5, 29.0, 29.0, 29.7, 29.7, 30.0, 30.0, 30.7, 31.0, 31.0, 
                31.5, 32.0, 32.0, 32.0, 33.0, 33.0, 33.5, 33.5, 34.0, 34.0, 34.5, 35.0, 
                35.0, 35.0, 35.0, 36.0, 36.0, 37.0, 38.5, 38.5, 39.5, 41.0, 41.0, 9.8, 
                10.5, 10.6, 11.0, 11.2, 11.3, 11.8, 11.8, 12.0, 12.2, 12.4, 13.0, 14.3, 15.0]
fish_weight = [242.0, 290.0, 340.0, 363.0, 430.0, 450.0, 500.0, 390.0, 450.0, 500.0, 475.0, 500.0, 
                500.0, 340.0, 600.0, 600.0, 700.0, 700.0, 610.0, 650.0, 575.0, 685.0, 620.0, 680.0, 
                700.0, 725.0, 720.0, 714.0, 850.0, 1000.0, 920.0, 955.0, 925.0, 975.0, 950.0, 6.7, 
                7.5, 7.0, 9.7, 9.8, 8.7, 10.0, 9.9, 9.8, 12.2, 13.4, 12.2, 19.7, 19.9]
```

이제 두 파이썬 리스트를 순회하면서 각 생선의 길이와 무게를 하나의 리스트로 담은 2차원 리스트를 만들겠다.

```python
# 각 생선의 길이와 무게를 하나의 리스트로 담은 2차원 리스트
fish_data = [[l, w] for l, w in zip(fish_length, fish_weight)]
fish_target = [1] * 35 + [0] * 14
```

이때 하나의 생선 데이터를 샘플 sample 이라고 부른다. 도미와 빙어는 각각 35마리, 14마리가 있으므로 전체 데이터는 49개의 샘플이 있다. 사용하는 특성은 길이와 무게 2개 이다. 이 데이터의 처음 35개를 훈련 세트로, 나머지 14개를 테스트 세트로 사용하겠다.

![image](https://github.com/user-attachments/assets/3fd7eb89-2d0b-4903-aa8c-06db207097f5)

KNeighborsClassifier 클래스를 임포트하고 모델 객체를 만들자

이제 전체 데이터에서 처음 35개를 선택해야 한다. 일반적으로 리스트처럼 배열의 요소를 선택할 때는 배열의 위치, 즉 인덱스(index)를 지정한다. 

```python
# 훈련 세트로 입력값 중 0부터 34번째 인덱스까지 사용
train_input = fish_data[:35]

# 훈련 세트로 타깃값 중 0부터 34번째 인덱스까지 사용
train_target = fish_data[:35]

# 테스트 세트로 입력값 중 35번째 부터 마지막 인덱스까지 사용
test_input = fish_data[35:]

# 테스트 세트로 타깃값 중 35번째 부터 마지막 인덱스까지 사용
test_target = fish_data[35:]
```

슬라이싱 연산으로 인덱스 0~34 까지 처음 35개 샘플을 훈련 세트로 선택했고, 인덱스 35~48까지 나머지 14개 샘플을 테스트 세트로 선택했다. 데이터를 준비했으니 훈련 세트로 fit() 메서드를 호출해 모델을 훈련하고, 테스트 세트로 score() 메서드를 호출해 평가해 보겠다.

```python
kn.fit(train_input, train_target)
kn.score(test_input, test_target)
```

![image](https://github.com/user-attachments/assets/5330d27b-861c-4e5d-8011-6ac9a69a6519)

정확도가 최악의 성능인 0.0을 나타낸다 무엇이 문제였을까?

**Comment by wongakim-99** [comment]
## 샘플링 편향

위에서 정확도가 0.0으로 나온 이유는 마지막 14개를 테스트 세트로 떼어 놓으면 훈련 세트에는 빙어가 하나도 들어있지 않는다. 즉, **빙어 없이 모델을 훈련하니 당연히 빙어를 올바르게 분류할 수 없던 것이다!!** 

이를 예방하기 위해서는 훈련 세트와 테스트 세트를 나누려면 도미와 빙어가 골고루 섞이게 만들어야 한다.

![image](https://github.com/user-attachments/assets/52ee36ce-ce84-417b-b9c3-785349c422c9)

상식적으로 훈련하는 데이터와 테스트하는 데이터에는 도미와 빙어가 골고루 섞여 있어야 한다. 일반적으로 훈련 세트와 테스트 세트에 샘플이 골고루 섞여 있지 않으면 샘플링이 한쪽으로 치우쳤다는 의미로 샘플링 편향(sampling bias)이라고 부른다.

위의 예시에서는 훈련 세트에 도미만 있기 때문에 테스트 세트가 무엇이든 무조건 도미라고 분류한다. 그런데 테스트 세트는 빙어만 있기 때문에 정답을 하나도 맞히지 못한다.

**Comment by wongakim-99** [comment]
## 넘파이

넘파이 numpy는 파이썬의 대표적인 배열 array 라이브러리이다. 앞서 파이썬의 리스트로 2차원 리스트를 표현할 수 있지만 고차원 리스트를 표현하려면 매우 번거롭다. 넘파이는 고차원의 배열을 손쉽게 만들고 조작할 수 있는 간편한 도구를 많이 제공한다.

1차원 배열은 선이고, 2차원 배열은 면, 3차원 배열은 3차원 공간을 나타낸다. 차원을 축으로 표시하면 다음과 같다.

![image](https://github.com/user-attachments/assets/8ee767e2-58ac-450a-a14e-f7c565b5b39b)

보통의 xy 좌표계와는 달리 시작점이 왼쪽 아래가 아니고 왼쪽 위에서부터 시작한다. 배열의 시작점을 이렇게 놓으면 편리한 점이 많다. 그럼 생선 데이터를 2차원 넘파일 배열로 변환해 보겠다.

```python
input_arr = np.array(fish_data)
target_arr = np.array(fish_target)
```

![image](https://github.com/user-attachments/assets/37c72828-2a76-4fe7-8468-7d0008f59ec7)

넘파이는 친절하게 배열의 차원을 구분하기 쉽도록 행과 열을 가지런히 출력한다. 출력 결과에서 49개의 행과 2개의 열을 쉽게 확인할 수 있다.

![image](https://github.com/user-attachments/assets/b8f330df-34ae-4aa7-8417-81286e848439)

이제 생선 데이터를 넘파이 배열로 준비했으므로 이 배열에서 랜덤하게 샘플을 선택해 훈련 세트와 테스트 세트로 만들 차례이다. 여기에서는 배열을 섞은 후에 나누는 방식 대신에 무작위로 샘플을 고르는 방법을 사용하겠다.

한 가지 주의할 점은 input_arr와 target_arr에서 같은 위치는 함께 선택되어야 한다는 점이다.
예를 들어 input_arr의 두 번째 값은 훈련 세트로 가고, target_arr의 두 번째 값은 테스트 세트로 가면 안된다. 타깃이 샘플과 함께 이동하지 않으면 올바르게 훈련될 수 없다.

![image](https://github.com/user-attachments/assets/5f7de4e4-50f9-4887-bce2-fba454a84e0d)

**Comment by wongakim-99** [comment]
## 두 번째 머신러닝 프로그램

앞서 만든 훈련 세트와 테스트 세트로 KNN 이웃 모델을 훈련시켜 보자. fit() 메서드를 실행 할 때마다 KNeighborsClassifier() 클래스의 객체는 이전에 학습한 모든 것을 잃어버린다. 이전 모델을 그대로 두고 싶다면 KNeighborsClassifier 클래스 객체를 새로 만들어야 한다. 여기서는 그냥 단순하게 이전에 만든 kn 객체를 그대로 사용하겠다.

```python
 kn.fit(train_input, train_target)
```

인덱스를 섞어 만든 train_input 과 train_target 으로 모델을 훈련시켰다. 다음은 test_input 과 test_target 으로 이 모델을 테스트할 차례이다.

```python
kn.fit(train_input, train_target)
kn.score(test_input, test_target)
```

결과는 1.0이 나왔다. 100%의 정확도로 테스트 세트에 있는 모든 생선을 맞혔다. predict() 메서드로 테스트 세트의 예측 결과와 실테 타깃을 확인해 보자.

![Image](https://github.com/user-attachments/assets/fb5dc996-e1ef-46e2-936c-5934f5f4d7af)

테스트 세트에 대한 예측 결과가 정답과 일치한다. predict() 메서드의 출력 결과가 test_target의 출력과 동일하게 array() 로 감싸 있는 것을 눈여겨 보자. 이 값은 넘파이 배열을 의미한다. 즉 predict() 메서드가 반환하는 값은 단순한 파이썬 리스트가 아니라 넘파이 배열이다. 사실 사이킷런 모델의 입력과 출력은 모두 넘파이 배열이다. 넘파이는 많은 파이썬 과학 라이브러리에서 사용하는 표준 데이터이다.

**Comment by wongakim-99** [comment]
## 훈련 모델 평가

뭔가 이상하다. 알고리즘이 도미와 빙어를 모두 외우고 있다면 같은 데이터로 모델을 평가하는 것은 이상하다. 모델을 훈련할 때 사용한 데이터로 모델의 성능을 평가하는 것은 정답을 미리 알려주고 시험을 보는 것과 같다. 공정하게 점수를 매기기 위해서는 훈련에 참여하지 않은 샘플을 사용해야 한다. 이 때문에 훈련 데이터를 훈련 세트와 테스트 세트로 나누었다. 훈련 세트로는 모델을 훈련하고 테스트 세트로 모델을 평가했다. 하지만 테스트 세트를 그냥 무작정 나누어서는 안된다. 도미와 빙어를 분류하는 것이 목적이기 때문에 훈련 세트나 테스트 세트에 어느 한 생선만 들어가 있다면 올바른 학습이 이루어지지 않을 것이다. 

도미와 빙어를 골고루 섞어 나누기 위해 파이썬의 다차원 배열 라이브러리인 넘파이를 사용해 보았다. 넘파이는 파이썬의 리스트와 비슷하지만 고차원의 큰 배열을 효과적으로 다룰 수 있고 다양한 도구를 많이 제공한다. 해당 절에서는 넘파이의 shuffle() 함수를 사용해 배열의 인덱스를 섞었다.

**Comment by wongakim-99** [comment]
# 데이터 전처리

올바른 결과 도출을 위해서 데이터를 사용하기 전에 데이터 전처리 과정을 거친다. 전처리 과정을 거친 데이터로 훈련했을 때의 차이를 알고 표준점수로 특성의 스케일을 변환하는 방법을 배우자.

---

## 넘파이로 데이터 준비하기

이번에는 조금 더 세련된 방법을 사용해 보겠다.

```python
fish_length = [25.4, 26.3, 26.5, 29.0, 29.0, 29.7, 29.7, 30.0, 30.0, 30.7, 31.0, 31.0, 
                31.5, 32.0, 32.0, 32.0, 33.0, 33.0, 33.5, 33.5, 34.0, 34.0, 34.5, 35.0, 
                35.0, 35.0, 35.0, 36.0, 36.0, 37.0, 38.5, 38.5, 39.5, 41.0, 41.0, 9.8, 
                10.5, 10.6, 11.0, 11.2, 11.3, 11.8, 11.8, 12.0, 12.2, 12.4, 13.0, 14.3, 15.0]
fish_weight = [242.0, 290.0, 340.0, 363.0, 430.0, 450.0, 500.0, 390.0, 450.0, 500.0, 475.0, 500.0, 
                500.0, 340.0, 600.0, 600.0, 700.0, 700.0, 610.0, 650.0, 575.0, 685.0, 620.0, 680.0, 
                700.0, 725.0, 720.0, 714.0, 850.0, 1000.0, 920.0, 955.0, 925.0, 975.0, 950.0, 6.7, 
                7.5, 7.0, 9.7, 9.8, 8.7, 10.0, 9.9, 9.8, 12.2, 13.4, 12.2, 19.7, 19.9]
```

이 전에는 이 파이썬 리스트를 순회하면서 원소를 하나씩 꺼내 생선 하나의 길이와 무게를 리스트 안의 리스트로 직접 구성했다. 하지만 이제 넘파이를 사용하여 훨씬 간단하게 만들어 보자.

넘파이를 우선 임포트하고 넘파이의 column_stack() 함수는 전달받은 리스트를 일렬로 세운 다음 차례대로 나란히 연결한다. 예를 들어 다음과 같은 간단한 2개의 리스트를 나란히 붙여 보겠다. 연결할 리스트는 파이썬 튜플로 전달한다.

```python
np.column_stack(([1,2,3], [4,5,6]))
```
<img width="195" alt="image" src="https://github.com/user-attachments/assets/934bf368-a5f9-404f-930c-a1f635cb8f73" />

[1,2,3], [4,5,6] 두 리스트를 일렬로 세운 다음 나란히 옆으로 붙였다. 만들어진 배열은 (3,2) 크기의 배열이다. 즉 3개의 행이 있고 2개의 열이 있다.

이제 fish_length 와 fish_weight 를 합치겠다. 방법은 동일하다.

```python
fish_data = np.column_stack((fish_length, fish_weight))
```
(print문으로 찍어서 확인하면 잘 연결된것을 확인할 수 있다.)
